File: AlgortihmEOS/Graph/5.py
n = int(input())

if n % 2 == 1:
    print(-1)
    exit()

ans = [0]
val = [1] * n
vis = [False] * n
adj_list = [[] for i in range(n)]

def dfs(node, a):
    vis[node] = True

    for adj in adj_list[node]:
        if vis[adj]: continue

        dfs(adj, a)

        if val[adj] % 2 == 1:
            val[node] += val[adj]
        else:
            a[0] += 1
 
for i in range(n-1):
    x, y = map(int, input().split())
    adj_list[x-1].append(y-1)
    adj_list[y-1].append(x-1)


dfs(0, ans)

print(ans[0])

File: AlgortihmEOS/Graph/5.cpp
#include <iostream>
#include <vector>

using namespace std;

int ans = 0;
vector<int> val;
vector<bool> vis;
vector<vector<int>> adj_list;

void dfs(int node);

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr); cout.tie(nullptr);

    int n;
    cin >> n;

    if (n % 2 == 1) {
        cout << -1 << endl;
        return 0;
    }

    val.resize(n, 1);
    vis.resize(n, false);
    adj_list.resize(n, vector<int>());

    int x, y;
    for(int i = 1; i < n; i++) {
        cin >> x >> y;
        x--; y--;
        adj_list[x].push_back(y);
        adj_list[y].push_back(x);
    }

    dfs(0);

    cout << ans;
}

void dfs(int node) {
    vis[node] = true;

    for(int adj : adj_list[node]) {
        if (vis[adj]) continue;

        dfs(adj);

        int value = val[adj];

        if (value & 1) {
            val[node] += value;
        } else {
            ans++;
        }
    }
}